File: scripts/bootstrap-plane.py
"""Provision one fresh local Plane instance from private Switchyard metadata.

Run through `python manage.py shell -c` in the pinned Plane v1.4.2 API image.
This follows plane/license/api/views/admin.py and plane/app/views/project/base.py;
keep this small direct-model integration covered when updating the Plane image.
"""

import json
import re
import sys
from uuid import UUID

from django.contrib.auth.hashers import make_password
from django.core.validators import validate_email
from django.db import transaction
from django.utils import timezone

from plane.db.models import (
    APIToken, Label, Profile, Project, ProjectIdentifier, ProjectMember,
    State, User, Workspace, WorkspaceMember,
)
from plane.db.models.state import DEFAULT_STATES
from plane.license.models import Instance, InstanceAdmin, InstanceConfiguration
from plane.utils.cache import invalidate_cache_directly


def require(condition, message):
    if not condition:
        raise SystemExit(message)


def create_project(workspace, owner, automation, project_id, identifier, name):
    project = Project.objects.create(id=project_id, workspace=workspace, name=name,
                                     identifier=identifier, network=0)
    ProjectIdentifier.objects.create(project=project, workspace=workspace, name=identifier)
    for user, role in ((owner, 20), (automation, 15)):
        ProjectMember.objects.create(project=project, workspace=workspace, member=user, role=role)
    states = DEFAULT_STATES + [
        {"name": "Human Review", "color": "#8E4EC6", "sequence": 37500, "group": "started"},
        {"name": "Blocked", "color": "#E5484D", "sequence": 40000, "group": "started"},
    ]
    State.objects.bulk_create([State(project=project, workspace=workspace, **state) for state in states])
    project.default_state = State.objects.get(project=project, default=True)
    project.save(update_fields=["default_state"])
    Label.objects.create(project=project, workspace=workspace, name="agent", color="#3E63DD")
    return project


def bootstrap():
    raw = json.load(sys.stdin)
    request = None
    if isinstance(raw, dict) and "board" in raw:
        require(set(raw) == {"board", "project"}, "Invalid project setup request.")
        data, request = raw["board"], raw["project"]
        require(isinstance(request, dict) and set(request) == {"id", "identifier", "name", "existing"}
                and all(isinstance(request[key], str) for key in ("id", "identifier", "name"))
                and isinstance(request["existing"], bool), "Invalid project setup request.")
        UUID(request["id"])
        require(re.fullmatch(r"[A-Z][A-Z0-9_]{0,11}", request["identifier"])
                and 1 <= len(request["name"]) <= 255, "Invalid project name or identifier.")
    else:
        data = raw
    fields = {
        "workspace", "workspace_id", "project_id", "identifier", "owner_email",
        "owner_id", "owner_password", "automation_id", "api_key",
    }
    require(isinstance(data, dict) and set(data) == fields
            and all(isinstance(value, str) for value in data.values()),
            "Invalid local Plane setup metadata.")
    ids = [UUID(data[field]) for field in
           ("workspace_id", "project_id", "owner_id", "automation_id")]
    require(len(set(ids)) == len(ids), "Local Plane identities must be distinct.")
    require(re.fullmatch(r"[a-z0-9][a-z0-9_-]{0,47}", data["workspace"]),
            "Invalid local Plane workspace slug.")
    require(re.fullmatch(r"[A-Z][A-Z0-9_]{0,11}", data["identifier"]),
            "Invalid local Plane project identifier.")
    validate_email(data["owner_email"])
    require(data["owner_email"] == data["owner_email"].lower()
            and data["owner_email"] != "symphony@switchyard.local",
            "Invalid local Plane owner email.")
    require(16 <= len(data["owner_password"]) <= 1024,
            "Invalid local Plane owner password.")
    require(re.fullmatch(r"plane_api_(?:[0-9a-f]{32}|[0-9a-f]{64})", data["api_key"]),
            "Invalid local Plane automation key.")

    with transaction.atomic():
        # The API entrypoint registers this row before it starts serving requests.
        # Upstream initial signup locks the same row to serialize registration.
        instance = Instance.objects.select_for_update().first()
        require(instance is not None, "Plane instance registration is incomplete.")
        owner = User.objects.filter(pk=data["owner_id"]).first()
        if owner is not None:
            # A repeat verifies ownership without resetting passwords, memberships,
            # or a revoked token. Later user-created projects and accounts are fine.
            automation = User.objects.filter(pk=data["automation_id"],
                email="symphony@switchyard.local", is_active=True,
                is_superuser=False, is_staff=False).first()
            valid = (
                instance.is_setup_done
                and owner.email == data["owner_email"] and owner.is_active
                and not owner.is_bot
                and InstanceAdmin.objects.filter(instance=instance, user=owner, role=20).exists()
                and Workspace.objects.filter(pk=data["workspace_id"],
                                             slug=data["workspace"], owner=owner).exists()
                and automation is not None
                and not InstanceAdmin.objects.filter(user_id=data["automation_id"]).exists()
            )
            for user_id, role in ((data["owner_id"], 20), (data["automation_id"], 15)):
                valid = valid and WorkspaceMember.objects.filter(
                    workspace_id=data["workspace_id"], member_id=user_id,
                    role=role, is_active=True).exists()
                if request is None:
                    valid = valid and ProjectMember.objects.filter(
                        project_id=data["project_id"], workspace_id=data["workspace_id"],
                        member_id=user_id, role=role, is_active=True).exists()
            if request is None:
                valid = valid and Project.objects.filter(pk=data["project_id"],
                    workspace_id=data["workspace_id"], identifier=data["identifier"]).exists()
            token = APIToken.objects.filter(token=data["api_key"],
                                            user_id=data["automation_id"],
                                            workspace_id=data["workspace_id"],
                                            is_active=True).first()
            valid = valid and token is not None and (
                token.expired_at is None or token.expired_at > timezone.now())
            # Earlier installations used a regular service user with no usable
            # password and a personal token. Preserve that exact non-admin
            # identity on restore; fresh setup still creates the native bot form.
            native = automation is not None and token is not None and (
                automation.is_bot and token.user_type == 1 and len(data["api_key"]) == 74)
            legacy = automation is not None and token is not None and (
                not automation.is_bot and not automation.has_usable_password()
                and token.user_type == 0 and len(data["api_key"]) == 42)
            valid = valid and (native or legacy)
            require(valid, "Local Plane setup no longer matches this installation; no changes made.")
        else:
            require(re.fullmatch(r"plane_api_[0-9a-f]{64}", data["api_key"]),
                    "Fresh local Plane setup requires a new automation key.")
            require(not instance.is_setup_done and not InstanceAdmin.objects.exists()
                    and not User.objects.exists() and not Workspace.objects.exists()
                    and not Project.objects.exists(),
                    "Plane already contains another setup; automatic setup requires a fresh instance.")
            owner = User.objects.create(
                id=data["owner_id"], username=UUID(data["owner_id"]).hex,
                email=data["owner_email"], password=make_password(data["owner_password"]),
                first_name="Switchyard", last_name="Owner", display_name="Switchyard Owner",
                is_password_autoset=False, is_email_verified=True,
            )
            InstanceAdmin.objects.create(instance=instance, user=owner, role=20)
            workspace = Workspace.objects.create(
                id=data["workspace_id"], name="Switchyard", slug=data["workspace"], owner=owner,
            )
            automation = User.objects.create(
                id=data["automation_id"], username=UUID(data["automation_id"]).hex,
                email="symphony@switchyard.local", password=make_password(None),
                display_name="Symphony", is_bot=True,
            )
            for user, role in ((owner, 20), (automation, 15)):
                WorkspaceMember.objects.create(workspace=workspace, member=user, role=role)
            create_project(workspace, owner, automation, data["project_id"], data["identifier"], "Tasks")
            Profile.objects.create(
                user=owner, is_onboarded=True, is_tour_completed=True,
                last_workspace_id=workspace.id,
                onboarding_step={"profile_complete": True, "workspace_create": True,
                                 "workspace_invite": True, "workspace_join": True},
            )
            APIToken.objects.create(
                user=automation, workspace=workspace, token=data["api_key"], user_type=1,
                label="Switchyard", allowed_rate_limit="600/minute",
            )
            for key, value in (("ENABLE_SIGNUP", "0"), ("ENABLE_EMAIL_PASSWORD", "1"),
                               ("ENABLE_MAGIC_LINK_LOGIN", "0")):
                InstanceConfiguration.objects.update_or_create(
                    key=key, defaults={"value": value, "category": "AUTHENTICATION",
                                       "is_encrypted": False},
                )
            instance.is_setup_done = True
            instance.save(update_fields=["is_setup_done"])

        if request is not None:
            workspace = Workspace.objects.get(pk=data["workspace_id"])
            automation = User.objects.get(pk=data["automation_id"])
            project = Project.objects.filter(pk=request["id"]).first()
            if project is None:
                require(not request["existing"], "The requested project does not exist in this workspace.")
                require(not ProjectIdentifier.objects.filter(workspace=workspace, name=request["identifier"]).exists(),
                        "Project identifier already exists; choose another --identifier.")
                create_project(workspace, owner, automation, request["id"], request["identifier"], request["name"])
            else:
                require(project.workspace_id == workspace.id and project.identifier == request["identifier"],
                        "Project identity does not match this workspace and identifier; no changes made.")
                if request["existing"]:
                    require(ProjectMember.objects.filter(project=project, member=owner, role=20, is_active=True).exists(),
                            "The local board owner must administer the requested project.")
                    membership, created = ProjectMember.objects.get_or_create(project=project, workspace=workspace,
                        member=automation, defaults={"role": 15})
                    require(created or (membership.role == 15 and membership.is_active),
                            "Existing automation project membership differs; no changes made.")
                    # Existing projects retain their states; add only missing workflow conventions.
                    for name, color, sequence in (("Human Review", "#8E4EC6", 37500), ("Blocked", "#E5484D", 40000)):
                        State.objects.get_or_create(project=project, workspace=workspace, name=name,
                            defaults={"color": color, "sequence": sequence, "group": "started"})
                    Label.objects.get_or_create(project=project, workspace=workspace, name="agent", defaults={"color": "#3E63DD"})
                else:
                    for user, role in ((owner, 20), (automation, 15)):
                        require(ProjectMember.objects.filter(project=project, workspace=workspace,
                            member=user, role=role, is_active=True).exists(),
                            "Existing project membership differs; no changes made.")

    # Readiness may have cached the pre-registration response for two hours.
    invalidate_cache_directly(path="/api/instances/", user=False)


try:
    bootstrap()
except Exception:
    # Never let a database error echo the JSON credentials or SQL parameters.
    raise SystemExit("Local Plane setup failed; retained private metadata can be used to retry.") from None

File: core/bootstrap_plane.go
package core

import (
	"bytes"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
)

type localPlane struct {
	Workspace    string `json:"workspace"`
	WorkspaceID  string `json:"workspace_id"`
	ProjectID    string `json:"project_id"`
	Identifier   string `json:"identifier"`
	OwnerID      string `json:"owner_id"`
	AutomationID string `json:"automation_id"`
	OwnerEmail   string `json:"owner_email"`
	Password     string `json:"owner_password"`
	APIKey       string `json:"api_key"`
}

func (a Installation) bootstrapPlane() (localPlane, error) {
	path := filepath.Join(a.Root, "plane.json")
	if _, err := os.Lstat(path); errors.Is(err, os.ErrNotExist) {
		settings := localPlane{Workspace: "switchyard", Identifier: "APP", OwnerEmail: "owner@switchyard.local"}
		for _, id := range []*string{&settings.WorkspaceID, &settings.ProjectID, &settings.OwnerID, &settings.AutomationID} {
			var value [16]byte
			if _, err := rand.Read(value[:]); err != nil {
				return localPlane{}, err
			}
			value[6] = value[6]&0x0f | 0x40
			value[8] = value[8]&0x3f | 0x80
			*id = fmt.Sprintf("%x-%x-%x-%x-%x", value[:4], value[4:6], value[6:8], value[8:10], value[10:])
		}
		for _, secret := range []*string{&settings.Password, &settings.APIKey} {
			var value [32]byte
			if _, err := rand.Read(value[:]); err != nil {
				return localPlane{}, err
			}
			*secret = hex.EncodeToString(value[:])
		}
		settings.APIKey = "plane_api_" + settings.APIKey
		data, _ := json.MarshalIndent(settings, "", "  ")
		// Persist the generated identity before provisioning so interrupted setup can resume.
		if err := writeOnce(path, append(data, '\n'), 0600); err != nil {
			return localPlane{}, err
		}
	} else if err != nil {
		return localPlane{}, err
	}
	settings, err := a.readLocalPlane()
	if err != nil {
		return localPlane{}, err
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return localPlane{}, err
	}
	if err := a.runPlaneSetup(data); err != nil {
		return localPlane{}, err
	}
	return settings, nil
}

func (a Installation) readLocalPlane() (localPlane, error) {
	path := filepath.Join(a.Root, "plane.json")
	info, err := os.Lstat(path)
	if err != nil {
		return localPlane{}, err
	}
	if !info.Mode().IsRegular() || info.Mode().Perm()&0077 != 0 {
		return localPlane{}, errors.New("plane.json must be a regular private file (chmod 600); existing credentials are preserved")
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return localPlane{}, err
	}
	var settings localPlane
	if err := json.Unmarshal(data, &settings); err != nil {
		return localPlane{}, errors.New("invalid plane.json; restore the original file before retrying")
	}
	return settings, nil
}

func (a Installation) runPlaneSetup(data []byte) error {
	code, err := os.ReadFile(filepath.Join(a.Root, "scripts/bootstrap-plane.py"))
	if err != nil {
		return err
	}
	command := a.command(filepath.Join(a.Root, "scripts/plane"), "exec", "-T", "api", "python", "manage.py", "shell", "-c", string(code))
	command.Stdin = bytes.NewReader(data)
	command.Stdout, command.Stderr = os.Stdout, os.Stderr
	if err := command.Run(); err != nil {
		return fmt.Errorf("local Plane setup failed; fix the reported cause and rerun the command (plane.json is preserved): %w", err)
	}
	return nil
}

func (a Installation) provisionProject(plane localPlane, options *ProjectOptions) error {
	existing := options.Project != ""
	if !existing {
		var err error
		options.Project, err = projectUUID(plane.WorkspaceID, options.Repo)
		if err != nil {
			return err
		}
	}
	if options.Identifier == "" {
		options.Identifier = projectIdentifier(options.Repo)
	}
	request := struct {
		Board   localPlane `json:"board"`
		Project struct {
			ID         string `json:"id"`
			Identifier string `json:"identifier"`
			Name       string `json:"name"`
			Existing   bool   `json:"existing"`
		} `json:"project"`
	}{Board: plane}
	request.Project.ID, request.Project.Identifier = options.Project, options.Identifier
	request.Project.Name = repositoryName(options.Repo)
	request.Project.Existing = existing
	data, err := json.Marshal(request)
	if err != nil {
		return err
	}
	return a.runPlaneSetup(data)
}
